agent/xds/listeners.go

```diff
@@ -71,16 +73,97 @@ func (s *Server) listenersFromSnapshotConnectProxy(cfgSnap *proxycfg.ConfigSnaps
 		}
 		resources[i+1] = upstreamListener
 	}
+
+	cfgSnap.Proxy.Expose.Finalize(s.Logger)
+	paths := cfgSnap.Proxy.Expose.Paths
+
+	// Add service health checks to the list of paths to create listeners for if needed
+	if cfgSnap.Proxy.Expose.Checks {
+		for _, check := range s.CheckFetcher.ServiceHTTPChecks(cfgSnap.Proxy.DestinationServiceID) {
+			p, err := parseCheckPath(check)
+			if err != nil {
+				s.Logger.Printf("[WARN] envoy: failed to create listener for check '%s': %v", check.CheckID, err)
+				continue
+			}
+			paths = append(paths, p)
+		}
+	}
+
+	// Configure additional listener for exposed check paths
+	for _, path := range paths {
+		clusterName := LocalAppClusterName
+		if path.LocalPathPort != cfgSnap.Proxy.LocalServicePort {
+			clusterName = makeExposeClusterName(path.LocalPathPort)
+		}
+
+		l, err := s.makeExposedCheckListener(cfgSnap, clusterName, path.Path, path.Protocol, path.ListenerPort)
+		if err != nil {
+			return nil, err
+		}
+		resources = append(resources, l)
+	}
+
 	return resources, nil
 }
 
+func parseCheckPath(check structs.CheckType) (structs.ExposePath, error) {
+	grpcRE := regexp.MustCompile("(.*)((?::)(?:[0-9]+))(.*)$")
+	httpRE := regexp.MustCompile(`^(http[s]?://)(\[.*?\]|\[?[\w\-\.]+)(:\d+)?([^?]*)(\?.*)?$`)
```

Comment:
Parsing URL with regexp is a bit of a smell. Is there a reason why `url.Parse` doesn't work here?

Reply:
I just didn't know about it, haha. This would have saved me from a lot of regex fiddling 🙈. Will try it out